tests/ga/test_extension.py

```diff
@@ -110,10 +110,14 @@ def _is_extension_dir(path):
     def _assert_ext_handler_status(self, aggregate_status, expected_status, version, expected_ext_handler_count=0):
         self.assertIsNotNone(aggregate_status, "Aggregate status should not be None")
         handler_statuses = aggregate_status['aggregateStatus']['handlerAggregateStatus']
-        self.assertEqual(expected_ext_handler_count, len(handler_statuses))
-        for ext_handler_status in handler_statuses:
-            self.assertEquals(expected_status, ext_handler_status['status'])
-            self.assertEquals(version, ext_handler_status['handlerVersion'])
+        try:
+            self.assertEqual(expected_ext_handler_count, len(handler_statuses))
+            for ext_handler_status in handler_statuses:
+                self.assertEquals(expected_status, ext_handler_status['status'])
+                self.assertEquals(version, ext_handler_status['handlerVersion'])
+        except Exception as e:
+            print("All ExtensionHandlers: {0}".format(handler_statuses))
```

Comment:
shouldn't you force a test failure here?

> self.fail("...message...")

Reply:
I'm re-raising the Exception below. It would be the same thing. The only reason I added a try catch is to print the handler_Statuses incase we need the debugging info later. That is how I was able to initially figure out that Timeout was causing the race condition.

Reply:
The  test will Error rather than Fail.  Maybe I'm misunderstanding your test. Who would raise?

Reply:
In case of assertion failures, an AssertionError (or something similar) is raised. In that case I'm just printing the info and moving on with the execution. It should technically fail the test as if nothing happened.

Random example: 

```python
All ExtensionHandlers: [{u'status': u'Ready', u'code': 0, u'formattedMessage': {u'lang': u'en-US', u'message': u'Plugin enabled'}, u'runtimeSettingsStatus': {u'settingsStatus': {u'status': {u'status': u'success', u'code': 0, u'name': u'OSTCExtensions.ExampleHandlerLinux', u'formattedMessage': {u'lang': u'en-US', u'message': None}, u'configurationAppliedTime': None, u'operation': None}, u'version': 1.0, u'timestampUTC': u'2020-05-29T19:52:49Z'}, u'sequenceNumber': 0}, u'useExactVersion': True, u'handlerVersion': u'1.0.0', u'handlerName': u'OSTCExtensions.ExampleHandlerLinux'}, {u'status': u'Ready', u'code': 0, u'formattedMessage': {u'lang': u'en-US', u'message': u'Plugin enabled'}, u'runtimeSettingsStatus': {u'settingsStatus': {u'status': {u'status': u'success', u'code': 0, u'name': u'Microsoft.Powershell.ExampleExtension', u'formattedMessage': {u'lang': u'en-US', u'message': None}, u'configurationAppliedTime': None, u'operation': None}, u'version': 1.0, u'timestampUTC': u'2020-05-29T19:52:49Z'}, u'sequenceNumber': 0}, u'useExactVersion': True, u'handlerVersion': u'1.0.0', u'handlerName': u'Microsoft.Powershell.ExampleExtension'}, {u'status': u'Ready', u'code': 0, u'formattedMessage': {u'lang': u'en-US', u'message': u'Plugin enabled'}, u'runtimeSettingsStatus': {u'settingsStatus': {u'status': {u'status': u'success', u'code': 0, u'name': u'Microsoft.EnterpriseCloud.Monitoring.ExampleHandlerLinux', u'formattedMessage': {u'lang': u'en-US', u'message': None}, u'configurationAppliedTime': None, u'operation': None}, u'version': 1.0, u'timestampUTC': u'2020-05-29T19:52:49Z'}, u'sequenceNumber': 0}, u'useExactVersion': True, u'handlerVersion': u'1.0.0', u'handlerName': u'Microsoft.EnterpriseCloud.Monitoring.ExampleHandlerLinux'}, {u'status': u'Ready', u'code': 0, u'formattedMessage': {u'lang': u'en-US', u'message': u'Plugin enabled'}, u'runtimeSettingsStatus': {u'settingsStatus': {u'status': {u'status': u'success', u'code': 0, u'name': u'Microsoft.CPlat.Core.ExampleExtensionLinux', u'formattedMessage': {u'lang': u'en-US', u'message': None}, u'configurationAppliedTime': None, u'operation': None}, u'version': 1.0, u'timestampUTC': u'2020-05-29T19:52:49Z'}, u'sequenceNumber': 0}, u'useExactVersion': True, u'handlerVersion': u'1.0.0', u'handlerName': u'Microsoft.CPlat.Core.ExampleExtensionLinux'}, {u'status': u'Ready', u'code': 0, u'formattedMessage': {u'lang': u'en-US', u'message': u'Plugin enabled'}, u'runtimeSettingsStatus': {u'settingsStatus': {u'status': {u'status': u'success', u'code': 0, u'name': u'Microsoft.OSTCExtensions.Edp.ExampleExtensionLinuxInTest', u'formattedMessage': {u'lang': u'en-US', u'message': None}, u'configurationAppliedTime': None, u'operation': None}, u'version': 1.0, u'timestampUTC': u'2020-05-29T19:52:49Z'}, u'sequenceNumber': 0}, u'useExactVersion': True, u'handlerVersion': u'1.0.0', u'handlerName': u'Microsoft.OSTCExtensions.Edp.ExampleExtensionLinuxInTest'}]


5 != 100000

Expected :100000
Actual   :5
<Click to see difference>

Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/larohra/WALinuxAgent-larohra-fork/tests/ga/test_extension.py", line 147, in test_cleanup_leaves_installed_extensions
    version="1.0.0")
  File "/home/larohra/WALinuxAgent-larohra-fork/tests/ga/test_extension.py", line 114, in _assert_ext_handler_status
    self.assertEqual(100000, len(handler_statuses))
```

Reply:
You should add the debug info as part of the assert message:

```
debug_info = "All ExtensionHandlers: {0}".format(handler_statuses)
...
self.assertEqual(expected_ext_handler_count, len(handler_statuses), "The number of handlers is incorrect.\n{0}'.format(debug_info))
...
etc

```

Reply:
Ahh good point! (Duh moment for me lol). Updated! :)